pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginwatcher
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	watcherapi "k8s.io/kubernetes/pkg/kubelet/apis/pluginregistration/v1alpha"
+	v1beta1 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta1"
+	v1beta2 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta2"
+)
+
+func TestExamplePlugin(t *testing.T) {
+	flag.Set("alsologtostderr", fmt.Sprintf("%t", true))
+	var logLevel string
+	flag.StringVar(&logLevel, "logLevel", "4", "test")
+	flag.Lookup("v").Value.Set(logLevel)
+	socketDir, err := ioutil.TempDir("", "plugin_test")
+	require.NoError(t, err)
+	socketPath := socketDir + "/plugin.sock"
+	w := NewWatcher(socketDir)
+
+	testCases := []struct {
+		description string
+		returnErr   error
+	}{
+		{
+			description: "Successfully register plugin through inotify",
+			returnErr:   nil,
+		},
+		{
+			description: "Successfully register plugin through inotify after plugin restarts",
+			returnErr:   nil,
+		},
+		{
+			description: "Fails registration with conflicting plugin name",
+			returnErr:   fmt.Errorf("conflicting plugin name"),
+		},
+		{
+			description: "Successfully register plugin during initial traverse after plugin watcher restarts",
+			returnErr:   nil,
+		},
+		{
+			description: "Fails registration with conflicting plugin name during initial traverse after plugin watcher restarts",
+			returnErr:   fmt.Errorf("conflicting plugin name"),
+		},
+	}
+
+	callbackCount := 0
+	w.AddHandler("example-plugin-type", func(name string, versions []string, sockPath string) error {
+		require.True(t, callbackCount <= (len(testCases)-1))
+		glog.Infof("receives plugin watcher callback for test: %s\n", testCases[callbackCount].description)
+		require.Equal(t, "example-plugin", name, "Plugin name mismatched!!")
+		require.Equal(t, []string{"v1beta1", "v1beta2"}, versions, "Plugin version mismatched!!")
+		// Verifies the grpcServer is ready to serve services.
+		conn, err := grpc.Dial(socketPath, grpc.WithInsecure(), grpc.WithBlock(),
```

Comment:
Can't we use dial and either:
- ignore the `watcherapi.NewRegistrationClient`
- refactor it to not return a `watcherapi.NewRegistrationClient` and take a timeout